compiler/src/dotty/tools/dotc/transform/init/Cache.scala

```diff
@@ -0,0 +1,132 @@
+package dotty.tools.dotc
+package transform
+package init
+
+import core.*
+import Contexts.*
+
+import ast.tpd
+import tpd.Tree
+
+class Cache[Config, Res]:
+  import Cache.*
+
+  /** The cache for expression values from last iteration */
+  protected var last: ExprValueCache[Config, Res] =  Map.empty
+
+  /** The output cache for expression values
+   *
+   *  The output cache is computed based on the cache values `last` from the
+   *  last iteration.
+   *
+   *  Both `last` and `current` are required to make sure an encountered
+   *  expression is evaluated once in each iteration.
+   */
+  protected var current: ExprValueCache[Config, Res] = Map.empty
+
+  /** Whether the current heap is different from the last heap?
+   *
+   *  `changed == false` implies that the fixed point has been reached.
+   */
+  protected var changed: Boolean = false
```

Comment:
Is `changed` part of the cache? Is the heap now intended to be independent of the cache? Should `changed` and `hasChanged` be renamed to indicate that they are about the heap?

Reply:
Yes, `changed` is part of the cache, it's not related to the heap.